graylog2-server/src/main/java/org/graylog2/streams/StreamServiceImpl.java

```diff
@@ -194,23 +195,57 @@ public Stream load(String id) throws NotFoundException {
         final Map<String, List<StreamRule>> allStreamRules = streamRuleService.loadForStreamIds(streamIds);
 
         final ImmutableList.Builder<Stream> streams = ImmutableList.builder();
+
+        final Map<String, IndexSet> indexSets = indexSetsForStreams(results);
+
+        final Set<String> outputIds = results.stream()
+                .map(this::outputIdsForRawStream)
+                .flatMap(outputs -> outputs.stream().map(ObjectId::toHexString))
+                .collect(Collectors.toSet());
+
+        final Map<String, Output> outputsById = outputService.loadByIds(outputIds)
+                .stream()
+                .collect(Collectors.toMap(Output::getId, Function.identity()));
+
+
         for (DBObject o : results) {
             final ObjectId objectId = (ObjectId) o.get("_id");
             final String id = objectId.toHexString();
             final List<StreamRule> streamRules = allStreamRules.getOrDefault(id, Collections.emptyList());
             LOG.debug("Found {} rules for stream <{}>", streamRules.size(), id);
 
-            final Set<Output> outputs = loadOutputsForRawStream(o);
+            final Set<Output> outputs = outputIdsForRawStream(o)
+                    .stream()
+                    .map(ObjectId::toHexString)
+                    .map(outputsById::get)
+                    .collect(Collectors.toSet());
 
             @SuppressWarnings("unchecked")
             final Map<String, Object> fields = o.toMap();
 
-            streams.add(new StreamImpl(objectId, fields, streamRules, outputs, getIndexSet(o)));
+            final String indexSetId = (String)fields.get(StreamImpl.FIELD_INDEX_SET_ID);
+
+            streams.add(new StreamImpl(objectId, fields, streamRules, outputs, indexSets.get(indexSetId)));
         }
 
         return streams.build();
     }
 
+    private List<ObjectId> outputIdsForRawStream(DBObject o) {
+        final List<ObjectId> objectIds = (List<ObjectId>) o.get(StreamImpl.FIELD_OUTPUTS);
+        return objectIds == null ? Collections.emptyList() : objectIds;
+    }
+
+    private Map<String, IndexSet> indexSetsForStreams(List<DBObject> streamsQuery) {
+        final Set<String> indexSetIds = streamsQuery.stream()
+                .map(stream -> (String)stream.get(StreamImpl.FIELD_INDEX_SET_ID))
+                .filter(s -> !isNullOrEmpty(s))
+                .collect(Collectors.toSet());
```

Comment:
Maybe we shouldn't call this `streamsQuery`? 

```suggestion
    private Map<String, IndexSet> indexSetsForStreams(List<DBObject> streams) {
        final Set<String> indexSetIds = streams.stream()
                .map(stream -> (String)stream.get(StreamImpl.FIELD_INDEX_SET_ID))
                .filter(s -> !isNullOrEmpty(s))
                .collect(Collectors.toSet());
```

Reply:
Absolutely! ✅